retrievalmarket/impl/providerstates/provider_states.go

```diff
@@ -58,18 +59,68 @@ func firstSuccessfulUnseal(ctx context.Context, node rm.RetrievalProviderNode, p
 }
 
 // UnsealData unseals the piece containing data for retrieval as needed
-func UnsealData(ctx fsm.Context, environment ProviderDealEnvironment, deal rm.ProviderDealState) error {
-	reader, err := firstSuccessfulUnseal(ctx.Context(), environment.Node(), *deal.PieceInfo)
+func UnsealData(ctx fsm.Context, environment ProviderDealEnvironment, deal rm.ProviderDealState, ds dagstore.DAGStore) error {
+	pieceCID := deal.PieceInfo.PieceCID.String()
+
+	// Can the sharded DAG Store serve a retrieval for this Piece ?
+	// This will be false ONLY if the deal is no longer active because even if the unsealed
+	// file for a deal has been deleted, the dag store can serve the retrieval by unsealing the file again
+	// since it's been told how to do so by calling the unsealing function (this "teach how to fetch by unsealing" thing is done when
+	// the shard is first activated either when the storage deal is published or below when we realize we don't have an active shard).
+	b, err := ds.IsShardActive(pieceCID)
 	if err != nil {
-		return ctx.Trigger(rm.ProviderEventUnsealError, err)
+		return nil
+	}
+
+	var rb dagstore.ReadOnlyBlockStore
+	if !b {
+		// This should ideally NEVER happen because the shard will be activated when the storage deal is published
+		// and will be removed ONLY when the storage deal expires.
+
+		// However, if it does happen, we need to activate the shard for this piece in the DAG Store.
+		// This call will block till the shard is activated i.e. an unsealed copy is fetched and indexed.
+		// It is possible that the DAG Store already has the index for this piece from a previous activation
+		// in which case we wont have to index it again.
+		// The one problem with this API and it's usage with the `GetShardReadOnlyBlockstore` call below is that the
+		// DAG Store will never memoize the unsealed file and so each call to `GetShardReadOnlyBlockstore` will result into one more fetch
+		// of the unsealed file but that should probably be okay as it seems to be a (Save space) vs (save time) tradeoff.
+		if err := ds.ActivateShard(pieceCID, &unsealedFetcher{
+			environment.Node(),
+			*deal.PieceInfo,
+		}); err != nil {
+			return err
+		}
+		// TODO We should remove the shard when the storage deal expires but that code probably belongs to the storage markets.
+	}
+
+	// get a read only blockstore that can be used to serve random access for the piece data here.
+	rb, err = ds.GetShardReadOnlyBlockstore(pieceCID)
```

Comment:
It would be great if we could somehow call IsShardActive and ActivateShard from within GetShardReadOnlyBlockstore.
Otherwise it may be tricky to synchronize between different threads - eg two threads could end up unsealing the same data in parallel.

Reply:
@dirkmc I agree. The current design assumes there's only one Markets process.